### ISSUE TITLE
ArgsParser: Parsing of flags after positional arguments is not self-consistent

### ISSUE THREAD

**Issue by nico** [SerenityOS/serenity#22759]
This works:

    Build/lagom/bin/pdf --render out.png    ~/Downloads/0000/0000342.pdf --page 5

This doesn't:

    Build/lagom/bin/pdf --render out.png    ~/Downloads/0000/0000342.pdf --page 5 --dump-contents

It does work if I move the `--dump-contents` in front of the name of the input filename, or if I move `--page 5` to earlier.

Maybe we only allow one flag after positional args or something?

It'd be nice if an arbitrary number of flags would work after positional args.

@alimpfard

**Comment by vincent-rg** [comment]
Hi,

I think I've just got to the root of this issue. It appears to be related to the argument shifting function in the option parser.

When correcting this, I've run those:
`pdf --render rendered.png --page 2 --dump-contents Documents/pdf/complex.pdf`
`pdf --render rendered.png --page 2 Documents/pdf/complex.pdf --dump-contents`
`pdf --render rendered.png --dump-contents Documents/pdf/complex.pdf --page 2`
`pdf --render rendered.png Documents/pdf/complex.pdf --page 2 --dump-contents`
without issue.

This is my first contribution, so it's gonna take me while to go through the process. I first have to read contributing guidelines in details :)

@kleinesfilmroellchen I don't think having seen any but I didn't go through all the code base: is there unit tests somewhere?

**Comment by kleinesfilmroellchen** [comment]
Not sure why you mention me, but yes, we have extensive unit tests in Tests. This should have tests under LibCore as well.

**Comment by vincent-rg** [comment]
Sorry, I didn't really know who to ask this :) I saw that you set the "bug" label.

thank you for this information!

**Comment by vincent-rg** [comment]
Just to provide a clearer explanation:

In the "not working" example, arguments are:
`--render out.png ~/Downloads/0000/0000342.pdf --page 5 --dump-contents`

`ArgsParser` relies on `OptionParser` to find... options.
`OptionParser::getopt()` has to be called until it tells us there's no more options.

When looking for the second option, `OptionParser` skips the filename positional argument and then finds `page` option.
`m_arg_index` is now 3, `m_skipped_arguments` is 1 and `m_consumed_args` is 2.
Found option (and eventually its value) is swapped with skipped arguments. The point is to have all skipped arguments at the end. Those will be the positional ones.

Arguments are now:
`--render out.png --page 5 ~/Downloads/0000/0000342.pdf --dump-contents`

Before returning the parsed option, `m_arg_index` is incremented by `m_consumed_args` and equals now 5.

`OptionParser::getopt()` is called again by `ArgsParser`. It returns the `dump-contents` option as it is at `m_arg_index` position (5).
There's no argument shifting, we're at the end.

Then, `ArgsParser.parse()` processes positional arguments ie it reads arguments after the last option.
With a bit of luck, as `ArgsParser` has himself counted consumed arguments (5 here), it detects that there's one argument left.
Good for him, it should have one and only one positional argument (the filename). Everything seems fine.
`ArgsParser` knows that positional argument is at the end because they are shifted, so it basically accesses "argument[5]", which is... `--dump-contents`.


TL;DR:
Simple wrong index error.

Fix: on arguments swap, update current argument index accordingly by taking into account skipped ones.

**Comment by nico** [comment]
@vincent-rg Thank you so much for fixing this! This has been really bothering me. I use `pdf` to test pdf changes on lagom often, and I ran into this multiple times a week for a long time now. No longer, finally :)
